datafusion/physical-expr/src/window/window_frame_state.rs

```diff
@@ -0,0 +1,972 @@
+            low,
+            min(high, data_size),
+        )?;
+        if low == data_size {
+            return Ok(None);
+        }
+        let val = range_columns
+            .iter()
+            .map(|arr| ScalarValue::try_from_array(arr, low))
+            .collect::<Result<Vec<ScalarValue>>>()?;
+        Ok(Some((val, low)))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use arrow::array::Float64Array;
+    use datafusion_common::ScalarValue;
+    use std::sync::Arc;
+
+    use crate::from_slice::FromSlice;
+
+    use super::*;
+
+    #[test]
+    fn test_find_next_group_and_start_index() {
+        const NUM_ROWS: usize = 6;
+        const NEXT_INDICES: [usize; 5] = [1, 2, 4, 4, 5];
+
+        let arrays: Vec<ArrayRef> = vec![
+            Arc::new(Float64Array::from_slice([5.0, 7.0, 8.0, 8., 9., 10.])),
+            Arc::new(Float64Array::from_slice([2.0, 3.0, 3.0, 3., 4.0, 5.0])),
+            Arc::new(Float64Array::from_slice([5.0, 7.0, 8.0, 8., 10., 11.0])),
+            Arc::new(Float64Array::from_slice([15.0, 13.0, 8.0, 8., 5., 0.0])),
+        ];
+        for (current_idx, next_idx) in NEXT_INDICES.iter().enumerate() {
+            let current_row_values = arrays
+                .iter()
+                .map(|col| ScalarValue::try_from_array(col, current_idx))
+                .collect::<Result<Vec<ScalarValue>>>()
+                .unwrap();
+            let next_row_values = arrays
+                .iter()
+                .map(|col| ScalarValue::try_from_array(col, *next_idx))
+                .collect::<Result<Vec<ScalarValue>>>()
+                .unwrap();
+            let res = WindowFrameStateGroups::find_next_group_and_start_index(
+                &arrays,
+                &current_row_values,
+                current_idx,
+            )
+            .unwrap();
+            assert_eq!(res, Some((next_row_values, *next_idx)));
+        }
+        let current_idx = NUM_ROWS - 1;
+        let current_row_values = arrays
+            .iter()
+            .map(|col| ScalarValue::try_from_array(col, current_idx))
+            .collect::<Result<Vec<ScalarValue>>>()
+            .unwrap();
+        let res = WindowFrameStateGroups::find_next_group_and_start_index(
+            &arrays,
+            &current_row_values,
+            current_idx,
+        )
+        .unwrap();
+        assert_eq!(res, None);
+    }
+
+    #[test]
+    fn test_window_frame_groups_preceding_huge_delta() {
+        const START: bool = true;
+        const END: bool = false;
+        const PRECEDING: bool = true;
+        const DELTA: u64 = 10;
+        const NUM_ROWS: usize = 5;
+
+        let arrays: Vec<ArrayRef> = vec![
+            Arc::new(Float64Array::from_slice([5.0, 7.0, 8.0, 9., 10.])),
+            Arc::new(Float64Array::from_slice([2.0, 3.0, 3.0, 4.0, 5.0])),
+            Arc::new(Float64Array::from_slice([5.0, 7.0, 8.0, 10., 11.0])),
+            Arc::new(Float64Array::from_slice([15.0, 13.0, 8.0, 5., 0.0])),
+        ];
```

Comment:
Stylistically, you can probably avoid a non trivial amount of repetition from these tests by refactoring the array functions into an individual function.